### PR TITLE
Use platform agnostic relative paths for workspace file (fix #148492)

### LINKED ISSUE #148492: Use platform agnostic relative paths for workspace file
Issue Type: <b>Bug</b>

1. Create a workspace file (`.code-workspace`) on a Windows machine.
2. Add the file to the git repository so the rest of the team can take advantage of a preconfigured one.
3. Clone the repository on a Linux machine (supposedly another team member who is a Linux user instead of a Windows one).
4. VS Code will complain about being unable to locate some workspace folders.

image

## Expected Behavior

VS Code would automatically convert Window-styled paths into Unix-styled paths (internally only) and vice versa.

## Walkthrough

Duplicate the file and manually convert the paths.

## Windows Machine Details (which created the file)

VS Code version: Code 1.66.2 (dfd34e8260c270da74b5c2d86d61aee4b6d56977, 2022-04-11T07:46:01.075Z)
OS version: Windows_NT x64 10.0.22000
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz (8 x 1992)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.89GB (5.19GB free)|
|Process Argv|--crash-reporter-id 900e10ac-fc10-4c51-8225-4957ba5ee1d8|
|Screen Reader|no|
|VM|0%|
</details>

### PR DESCRIPTION
**Pull request by bpasero**
cc @aeschli 

This makes use of `/` for all relative paths even on Windows. We only preserve `\` for absolute paths and only if we are not yet using `/` (this was there before already).

### CODE REVIEW

**Review by sandy081 — DISMISSED** [review]

**Comment by bpasero** [comment]
@sandy081 sorry this actually caused unit test failures on Windows and that made me revisit the fix. It is a bit smaller now and it will make sure we still use backslash for absolute paths even when some relative paths use slash, because I still think backslash for absolute paths is correct on Windows and we should only use slash if other absolute paths use slash.

File: src/vs/platform/workspaces/common/workspaces.ts [hunk 1]
```diff
@@ -265,11 +260,13 @@ function doParseStoredWorkspace(path: URI, contents: string): IStoredWorkspace {
...
-	if (isWindows) {
-		return storedFolders.some(folder => isRawFileWorkspaceFolder(folder) && folder.path.indexOf('/') >= 0);
+	if (!isWindows) {
+		return true; // Linux/macOS always use slash
 	}
 
-	return true;
+	// Only prefer slashes if other absolute paths in the
+	// workspace configuration file use slashes as well
+	return storedFolders.some(folder => isRawFileWorkspaceFolder(folder) && isAbsolute(folder.path) && folder.path.indexOf(posix.sep) >= 0);
```
Comment (sandy081) [inline src/vs/platform/workspaces/common/workspaces.ts:269]: Not sure if I understand this. Comment says to use slash only if absolute folder paths use slashes but the condition reads: use slash if there is a folder with absolute path that uses backslash

**Review by sandy081 — CHANGES_REQUESTED** [review]

File: src/vs/platform/workspaces/common/workspaces.ts (hunk 1 above)
Comment (bpasero) [reply src/vs/platform/workspaces/common/workspaces.ts:269]: @sandy081 `posix.sep` is `/` also on Windows, can you recheck?

**Review by sandy081 — APPROVED** [review]

File: src/vs/platform/workspaces/common/workspaces.ts [hunk 2]
```diff
@@ -124,55 +124,77 @@ export interface IEnterWorkspaceResult {
...
 		}
 	}
 
 	return { name: folderName, path: folderPath };
 }
 
+function massagePathForWindows(folderPath: string) {
+
+	// Drive letter should be upper case
+	folderPath = normalizeDriveLetter(folderPath);
```
Comment (aeschli) [inline src/vs/platform/workspaces/common/workspaces.ts:186]: This could go inside the `!isUNC` then clause

**Pull request merged**